sapio-psbt/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright Judica, Inc 2022
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+//  License, v. 2.0. If a copy of the MPL was not distributed with this
+//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use bitcoin::consensus::serialize;
+use bitcoin::schnorr::TapTweak;
+use bitcoin::secp256k1::rand::Rng;
+use bitcoin::secp256k1::{rand, All};
+use bitcoin::util::bip32::{ExtendedPubKey, KeySource};
+use bitcoin::util::sighash::Prevouts;
+use bitcoin::util::taproot::TapLeafHash;
+use bitcoin::util::taproot::TapSighashHash;
+use bitcoin::XOnlyPublicKey;
+use bitcoin::{
+    psbt::PartiallySignedTransaction, secp256k1::Secp256k1, util::bip32::ExtendedPrivKey,
+};
+use bitcoin::{KeyPair, TxOut};
+use bitcoin::{Network, SchnorrSig};
+use miniscript::psbt::PsbtExt;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::error::Error;
+use std::fmt::Display;
+use std::str::FromStr;
+
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum PSBTApi {
+    Finished {
+        completed: bool,
+        hex: String,
+    },
+    NotFinished {
+        completed: bool,
+        psbt: String,
+        error: String,
+        errors: Vec<String>,
+    },
+}
+
+pub fn finalize_psbt_format_api(psbt: PartiallySignedTransaction) -> PSBTApi {
+    let secp = Secp256k1::new();
+    psbt.finalize(&secp)
+        .map(|tx| {
+            let hex = bitcoin::consensus::encode::serialize_hex(&tx.extract_tx());
+            PSBTApi::Finished {
+                completed: true,
+                hex,
+            }
+        })
+        .unwrap_or_else(|(psbt, errors)| {
+            let errors: Vec<_> = errors.iter().map(|e| format!("{:?}", e)).collect();
+            let encoded_psbt = base64::encode(serialize(&psbt));
+            PSBTApi::NotFinished {
+                completed: false,
+                psbt: encoded_psbt,
+                error: "Could not fully finalize psbt".into(),
+                errors,
+            }
+        })
+}
+
+pub async fn read_key_from_file(
+    file_name: &std::ffi::OsStr,
+) -> Result<ExtendedPrivKey, Box<dyn Error>> {
+    let buf = tokio::fs::read(file_name).await?;
+    Ok(ExtendedPrivKey::decode(&buf)?)
+}
+
+pub fn sign(
+    xpriv: ExtendedPrivKey,
+    mut psbt: PartiallySignedTransaction,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<Vec<u8>, PSBTSigningError> {
+    sign_psbt_inplace(&xpriv, &mut psbt, &Secp256k1::new(), hash_ty)?;
+    let bytes = serialize(&psbt);
+    Ok(bytes)
+}
+pub async fn show_pubkey(input: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let buf = tokio::fs::read(input).await?;
+    let xpriv = ExtendedPrivKey::decode(&buf)?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
```

Comment:
o.O

Reply:
fwiw this is a move-only change, this was previously just a helper in the cli module... probably can be cleaned up more separating library v.s. application code.